### PR TITLE
Refactor ProcessGroupNCCL collective primitives

### PR DESCRIPTION
**Pull request by mxw**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #19631 Fix CUDA stream syncing bug in allgather and reduce_scatter
* #18844 Add support for reduce-scatter in c10d
* **#18820 Refactor ProcessGroupNCCL collective primitives**

Summary:
Planning to add reduce-scatter, but no room in my stomach for more
copypasta.

Also rewrote the tensor list validation logic.  The existing validation
was ill-suited for all the cases it was being used for; it took a vector
of input tensors and a vector of output tensors, but only ever received
either two references to the same vector, or a bespoke singleton vector
and a vector of outputs (for which it would ignore all but the first
output).  In the first case, it performed unnecessary checks, and in the
second, it skipped necessary ones.

Differential Revision: [D14762369](https://our.internmc.facebook.com/intern/diff/D14762369)

### CODE REVIEW

File: torch/lib/c10d/ProcessGroupNCCL.cpp [hunk 1]
```diff
@@ -287,90 +287,107 @@ std::vector<std::shared_ptr<NCCLComm>>& ProcessGroupNCCL::getNCCLComm(
...
-        "The number of input tensors is larger than "
-        "the number of available GPUs");
+      "Tensor list mustn't be larger than the number of available GPUs"
+    );
   }
 
-  // To make sure each tensor is on separate devices
-  std::unordered_set<int> usedDevices;
-  usedDevices.reserve(input.size());
+  auto const& first = tensors.front();
```
Comment (pietern) [inline torch/lib/c10d/ProcessGroupNCCL.cpp]: Prefer `const auto` over `auto const`

```
$ git grep 'const auto' | wc -l
3782
$ git grep 'auto const' | wc -l
16
```

File: torch/lib/c10d/ProcessGroupNCCL.cpp [hunk 2]
```diff
@@ -287,90 +287,107 @@ std::vector<std::shared_ptr<NCCLComm>>& ProcessGroupNCCL::getNCCLComm(
...
+  if (tensors.size() == 0) {
+    throw std::runtime_error("Tensor list must be nonempty");
   }
-
-  if (input.size() > static_cast<size_t>(at::cuda::getNumGPUs())) {
+  if (tensors.size() > static_cast<size_t>(at::cuda::getNumGPUs())) {
     throw std::runtime_error(
-        "The number of input tensors is larger than "
-        "the number of available GPUs");
+      "Tensor list mustn't be larger than the number of available GPUs"
```
Comment (pietern) [inline torch/lib/c10d/ProcessGroupNCCL.cpp]: `clang-format` will reformat this with bigger indent

File: torch/lib/c10d/ProcessGroupNCCL.cpp [hunk 3]
```diff
@@ -287,90 +287,107 @@ std::vector<std::shared_ptr<NCCLComm>>& ProcessGroupNCCL::getNCCLComm(
...
-          "type");
+    if (t.scalar_type() != first.scalar_type()) {
+      throw std::runtime_error("Tensors must have identical type");
     }
-    // Check the input tensor size is identical
-    if (input[i].numel() != inputNumElement) {
-      throw std::runtime_error(
-          "Expecting all input tensors to have identical "
-          "number of elements");
+    if (t.numel() != first.numel()) {
```
Comment (pietern) [inline torch/lib/c10d/ProcessGroupNCCL.cpp]: Comparing `sizes` is even better, seeing as mixing different dimensions that happen to have the same number of elements is not a good idea.

**Review by pietern — COMMENTED** [review]
Nice work, this dedups quite a bit. There are some CI failures to check out though.

File: torch/lib/c10d/ProcessGroupNCCL.cpp (hunk 2 above)
Comment (mxw) [reply torch/lib/c10d/ProcessGroupNCCL.cpp]: Does PyTorch use clang-format as a source-of-truth?  I'm freaked out about what it does to lambdas.

Comment (mxw) [reply torch/lib/c10d/ProcessGroupNCCL.cpp]: This file is decidedly not clang-formatted, and running clang-format indeed turns the lambda formatting into spaghetti.  I'll leave it for now and defer to the code formatting zealots who surely will visit upon this file if they find it necessary at some point 😬

File: torch/lib/c10d/ProcessGroupNCCL.cpp [hunk 4]
```diff
@@ -287,90 +287,107 @@ std::vector<std::shared_ptr<NCCLComm>>& ProcessGroupNCCL::getNCCLComm(
...
-        "number of tensors as the output tensor sequence");
-  }
+namespace {
 
-  if (input.size() == 0) {
-    throw std::runtime_error("The number of input tensors should not be zero");
+/*
+ * Check that all `tensors' have the same type and shape and are distributed
+ * across distinct GPUs.
+ */
```
Comment (pietern) [inline torch/lib/c10d/ProcessGroupNCCL.cpp]: All comments in this directory use `//` style.

Did a quick check in the rest of the repository and the block comment is not prevalent elsewhere either:

```
$ grep -ir '^//' * | wc -l
3179
$ grep -ir '^/\*' * | wc -l
33
```

File: torch/lib/c10d/ProcessGroupNCCL.cpp (hunk 2 above)
Comment (pietern) [reply torch/lib/c10d/ProcessGroupNCCL.cpp]: There are a few other clang-format violations in c10d right now, so we can address this in a separate commit. Ideally I'd like to setup a clang-format linter for this directory (and others where applicable) so we don't have to deal with partially formatted files.

File: torch/lib/c10d/ProcessGroupNCCL.cpp [hunk 5]
```diff
@@ -287,90 +287,107 @@ std::vector<std::shared_ptr<NCCLComm>>& ProcessGroupNCCL::getNCCLComm(
...
 
-    bool inserted;
-    std::tie(std::ignore, inserted) = usedDevices.insert(input[i].get_device());
-    // Device verification, if the insertion didn't take place
-    if (!inserted) {
-      throw std::runtime_error("Expecting inputs on different GPU devices");
+/*
+ * Flatten each list in `tensor_lists' for a gather or scatter operation, and
+ * ensure compatibility with the corresponding tensor in `other'.
+ */
```
Comment (pietern) [inline torch/lib/c10d/ProcessGroupNCCL.cpp]: Comment style.

File: torch/lib/c10d/ProcessGroupNCCL.cpp [hunk 6]
```diff
@@ -287,90 +287,107 @@ std::vector<std::shared_ptr<NCCLComm>>& ProcessGroupNCCL::getNCCLComm(
...
+    throw std::runtime_error(
+      "Tensor list operands to scatter/gather must have the same length"
+    );
+  }
+  const auto num_devices = tensor_lists.size();
+
+  std::vector<at::Tensor> flattened;
+  flattened.resize(num_devices);
+
+  for (auto i = size_t{}; i < num_devices; ++i) {
```
Comment (pietern) [inline torch/lib/c10d/ProcessGroupNCCL.cpp]: I've never seen this before and have always used `size_t i`. TIL.

File: torch/lib/c10d/ProcessGroupNCCL.cpp [hunk 7]
```diff
@@ -384,235 +401,164 @@ std::shared_ptr<ProcessGroup::Work> ProcessGroupNCCL::allreduce(
...
 
-    // Input `tensors` are created on a worker stream and used in different
-    // ncclStream. Hence, `tensors` must record the ncclStream to prevent being
-    // freed before ncclAllReduce finishes. See [Sync Streams].
+    // Both `inputs' and `outputs' are created on a worker stream and used in
+    // different ncclStreams.  Hence, both must record the ncclStream to
+    // prevent being freed before the collective finishes.
+    //
+    // We only record `inputs' here, and leave recording `outputs' to `fn' for
+    // operations where `inputs' and `outputs' are the same.
```
Comment (pietern) [inline torch/lib/c10d/ProcessGroupNCCL.cpp]: "[...] where inputs and outputs are **NOT** the same.", otherwise it would be redundant, no?

**Review by pietern — COMMENTED** [review]
Some additional nits.

In general this looks good to me. Thanks for reducing the amount of copy/paste here!

I didn't fully vet the logic in every path and trust that our tests provide with enough coverage.

@mrshenli Can you also take a look?

File: torch/lib/c10d/ProcessGroupNCCL.cpp (hunk 7 above)
Comment (mxw) [reply torch/lib/c10d/ProcessGroupNCCL.cpp]: Yep—will fix.

**Comment by mxw** [comment]
Most recent update addresses all of @pietern's comments.

**Comment by mrshenli** [comment]
@mxw you might need a rebase

**Review by mrshenli — APPROVED** [review]
I think the lint failure is irrelevant.

```
57.63s$ flake8
./test/test_fake_quant.py:28:1: E302 expected 2 blank lines, found 1
The command "flake8" exited with 1.
```

@pietern do you have any comments to add?

**Pull request closed without merging**